### PR TITLE
[2.next] Add support for having/lock options

### LINKED ISSUE #9812: [2.next] Adding forUpdate and having for the Model::find() options.
This is a (multiple allowed):

* [ ] bug
* [x] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 2.10.0

I would like to use `SELECT FOR UPDATE` in CakePHP2. Without `SELECT FOR UPDATE`, applications would be easy to break. For example:
```php
// In transaction
$count = $this->Bookmark->findCountByUserId($this->Auth->user('id'));
if ($count < 10) {
    $this->Bookmark->save($this->request->data);
}
```
The above code doesn't prevent us from inserting more than 10 records actually.

I found [an old post about SELECT FOR UPDATE](http://bakery.cakephp.org/2012/04/14/SELECT-FOR-UPDATE-hack-kind-of-ugly-but-it-works-PostgreSQL-and-MySql.html)  in the Bakery, but it doesn't work in CakePHP2. If `Model::find()` supports for the `forUpdate` option, we can do it easy:
```php
// In transaction
$count = $this->Bookmark->find('count', [
    'conditions' => ['user_id' => $this->Auth->user('id')],
    'forUpdate' => true,
]);
if ($count < 10) {
    $this->Bookmark->save($this->request->data);
}
```
`forUpdate` will be mapped to a correct statement for each database drivers.

If people don't mind, I would like to also allow users to call it dynamically, like the first example:
```php
$count = $this->Bookmark->findCountByUserIdForUpdate($this->Auth->user('id));
```
The alternative option is to add `epilog`, like CakePHP3. Personally, I would like to add `forUpdate`, as it allows people to write non database-specific code. But if people prefer `epilog` over `forUpdate`, I take the opinion.

Adding `having` is not a main issue. But I can add it while I am implementing `forUpdate`.

### PR DESCRIPTION
**Pull request by chinpei215**
Fixes #9812 

I didn't add support for `LOCK IN SHARE MODE`/`FOR SHARE`/`WITH (SERIALIZABLE)` and dynamic finder, to make things simple. After this PR has been merged, I'll try to update related 2.next documents.

### CODE REVIEW

File: lib/Cake/Test/Case/Model/Datasource/Database/SqlserverTest.php [hunk 1]
```diff
@@ -707,4 +707,117 @@ public function testStoredProcedureReturn() {
...
+		// HAVING
+		$query = array(
+			'fields' => array('user_id', 'COUNT(*) AS count'),
+			'table' => 'articles',
+			'alias' => 'Article',
+			'group' => 'user_id',
+			'order' => array('COUNT(*)' => 'DESC'),
+			'limit' => 5,
+			'having' => array('COUNT(*) >' => 10),
+		);
```
Comment (markstory) [inline lib/Cake/Test/Case/Model/Datasource/Database/SqlserverTest.php]: Should having() be in a separate test method?

Comment (chinpei215) [inline lib/Cake/Test/Case/Model/Datasource/Database/SqlserverTest.php]: Do you mean I should separate `testBuildStatement` into two methods - `testBuildStatementWithHaving()` and `testBuildStatementWithLockingHint()` ?

Comment (markstory) [inline lib/Cake/Test/Case/Model/Datasource/Database/SqlserverTest.php]: Yes.

Comment (chinpei215) [inline lib/Cake/Test/Case/Model/Datasource/Database/SqlserverTest.php]: OK. I'll do that tomorrow.

**Comment by chinpei215** [comment]
Done.

File: lib/Cake/Model/Datasource/Database/Sqlite.php [hunk 2]
```diff
@@ -591,4 +591,14 @@ public function nestedTransactionSupported() {
 		return $this->useNestedTransactions && version_compare($this->getVersion(), '3.6.8', '>=');
 	}
 
+/**
+ * Returns a locking hint for the given mode.
+ * Sqlite Datasource doesn't support row-level locking.
```
Comment (ravage84) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: Separate with a blank line. So the first becomes the short description, the second the "long" description.

File: lib/Cake/Model/Datasource/Database/Sqlite.php [hunk 3]
```diff
@@ -591,4 +591,14 @@ public function nestedTransactionSupported() {
...
+ *
+ * @param mixed $mode Lock mode
+ * @return string|null
```
Comment (ravage84) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: May be add a return description. With an example, may be.

File: lib/Cake/Model/Datasource/Database/Sqlserver.php [hunk 4]
```diff
@@ -814,4 +817,17 @@ public function getSchemaName() {
 		return $this->config['schema'];
 	}
 
+/**
+ * Returns a locking hint for the given mode.
+ * Currently, this method only returns WITH (UPDLOCK) when the mode is true.
```
Comment (ravage84) [inline lib/Cake/Model/Datasource/Database/Sqlserver.php]: Same as above.

File: lib/Cake/Model/Datasource/Database/Sqlserver.php [hunk 5]
```diff
@@ -814,4 +817,17 @@ public function getSchemaName() {
...
+ *
+ * @param mixed $mode Lock mode
+ * @return string|null
```
Comment (ravage84) [inline lib/Cake/Model/Datasource/Database/Sqlserver.php]: Same as above.

File: lib/Cake/Model/Datasource/DboSource.php [hunk 6]
```diff
@@ -2041,7 +2047,9 @@ public function renderStatement($type, $data) {
 
 		switch (strtolower($type)) {
 			case 'select':
-				return trim("SELECT {$fields} FROM {$table} {$alias} {$joins} {$conditions} {$group} {$order} {$limit}");
+				$having = !empty($having) ? " $having" : '';
+				$lock = !empty($lock) ? " $lock" : '';
+				return trim("SELECT {$fields} FROM {$table} {$alias} {$joins} {$conditions} {$group}{$having} {$order} {$limit}{$lock}");
```
Comment (ravage84) [inline lib/Cake/Model/Datasource/DboSource.php]: Quite a long line. May be asemble a line above and then return for better better readability?
Alternatively you could use multiple lines.

File: lib/Cake/Model/Datasource/DboSource.php [hunk 7]
```diff
@@ -3126,6 +3136,35 @@ public function group($fields, Model $Model = null) {
 	}
 
 /**
+ * Create a HAVING SQL clause.
+ *
+ * @param mixed $fields Array or string of conditions
+ * @param bool $quoteValues If true, values should be quoted
+ * @param Model $Model A reference to the Model instance making the query
+ * @return string|null
```
Comment (ravage84) [inline lib/Cake/Model/Datasource/DboSource.php]: Return description missing.

File: lib/Cake/Model/Datasource/DboSource.php [hunk 8]
```diff
@@ -3126,6 +3136,35 @@ public function group($fields, Model $Model = null) {
...
+		}
+		return ' HAVING ' . $this->conditions($fields, $quoteValues, false, $Model);
+	}
+
+/**
+ * Returns a locking hint for the given mode.
+ * Currently, this method only returns FOR UPDATE when the mode is true.
+ *
+ * @param mixed $mode Lock mode
+ * @return string|null
```
Comment (ravage84) [inline lib/Cake/Model/Datasource/DboSource.php]: Return description missing.

File: lib/Cake/Model/Datasource/DboSource.php [hunk 9]
```diff
@@ -3126,6 +3136,35 @@ public function group($fields, Model $Model = null) {
...
+	public function having($fields, $quoteValues = true, Model $Model = null) {
+		if (!$fields) {
+			return null;
+		}
+		return ' HAVING ' . $this->conditions($fields, $quoteValues, false, $Model);
+	}
+
+/**
+ * Returns a locking hint for the given mode.
+ * Currently, this method only returns FOR UPDATE when the mode is true.
```
Comment (ravage84) [inline lib/Cake/Model/Datasource/DboSource.php]: Same as above.

File: lib/Cake/Test/Case/Model/Datasource/Database/SqlserverTest.php [hunk 10]
```diff
@@ -707,4 +707,139 @@ public function testStoredProcedureReturn() {
...
+
+		$sql = $db->buildStatement($query, $this->model);
+		$expected = 'SELECT  TOP 5 user_id, COUNT(*) AS count FROM articles AS [Article]   WHERE 1 = 1  GROUP BY user_id  HAVING COUNT(*) > 10  ORDER BY COUNT(*) DESC';
```
Comment (ravage84) [inline lib/Cake/Test/Case/Model/Datasource/Database/SqlserverTest.php]: Can you split that to multiple lines, so it's not so long?

File: lib/Cake/Test/Case/Model/Datasource/Database/SqlserverTest.php [hunk 11]
```diff
@@ -707,4 +707,139 @@ public function testStoredProcedureReturn() {
...
+		$this->assertEquals($expected, $sql);
+
+		$sql = $db->buildStatement(array('offset' => 15) + $query, $this->model);
+		$expected = 'SELECT user_id, COUNT(*) AS count FROM articles AS [Article]   WHERE 1 = 1  GROUP BY user_id  HAVING COUNT(*) > 10  ORDER BY COUNT(*) DESC  OFFSET 15 ROWS FETCH FIRST 5 ROWS ONLY';
```
Comment (ravage84) [inline lib/Cake/Test/Case/Model/Datasource/Database/SqlserverTest.php]: Same as abovve.

File: lib/Cake/Test/Case/Model/Datasource/Database/SqlserverTest.php [hunk 12]
```diff
@@ -707,4 +707,139 @@ public function testStoredProcedureReturn() {
...
+			'limit' => 1,
+			'lock' => true,
+		);
+
+		$sql = $db->buildStatement($query, $this->model);
+		$expected = 'SELECT  TOP 1 id FROM users AS [User]  WITH (UPDLOCK)   WHERE 1 = 1   ORDER BY [id] ASC';
+		$this->assertEquals($expected, $sql);
+
+		$sql = $db->buildStatement(array('offset' => 15) + $query, $this->model);
+		$expected = 'SELECT id FROM users AS [User]  WITH (UPDLOCK)   WHERE 1 = 1   ORDER BY [id] ASC  OFFSET 15 ROWS FETCH FIRST 1 ROWS ONLY';
```
Comment (ravage84) [inline lib/Cake/Test/Case/Model/Datasource/Database/SqlserverTest.php]: Same as agove.

File: lib/Cake/Test/Case/Model/Datasource/Database/SqlserverTest.php [hunk 13]
```diff
@@ -707,4 +707,139 @@ public function testStoredProcedureReturn() {
...
+			'table' => 'articles',
+			'alias' => 'Article',
+			'group' => 'user_id',
+			'order' => array('COUNT(*)' => 'DESC'),
+			'limit' => 5,
+			'having' => array('COUNT(*) >' => 10),
+		);
+
+		$sql = $db->buildStatement($query, $this->model);
+		$expected = 'SELECT  TOP 5 user_id, COUNT(*) AS count FROM articles AS [Article]   WHERE 1 = 1  GROUP BY user_id  HAVING COUNT(*) > 10  ORDER BY COUNT(*) DESC';
```
Comment (ravage84) [inline lib/Cake/Test/Case/Model/Datasource/Database/SqlserverTest.php]: Same as above.

File: lib/Cake/Test/Case/Model/Datasource/DboSourceTest.php [hunk 14]
```diff
@@ -1448,6 +1476,63 @@ public function testBuildStatementDefaults() {
...
+				'table' => 'articles',
+				'alias' => 'Article',
+				'group' => 'user_id',
+				'order' => array('COUNT(*)' => 'DESC'),
+				'limit' => 5,
+				'having' => array('COUNT(*) >' => 10),
+			),
+			$this->Model
+		);
+		$expected = 'SELECT user_id, COUNT(*) AS count FROM articles AS Article   WHERE 1 = 1  GROUP BY user_id  HAVING COUNT(*) > 10  ORDER BY COUNT(*) DESC  LIMIT 5';
```
Comment (ravage84) [inline lib/Cake/Test/Case/Model/Datasource/DboSourceTest.php]: Too long.

**Review by ravage84 — CHANGES_REQUESTED** [review]
Some nitpicking from my side.

**Comment by ravage84** [comment]
@chinpei215 while you are at it, could you add ``@covers`` and ``@coversDefaultClass`` annotations to these tests? This way, the coverage will not lie about what actually is covered by your tests. 😼

**Comment by chinpei215** [comment]
@ravage84 Thanks. I'll update them when I have time.

**Comment by chinpei215** [comment]
Added return description. But I didn't split the long lines to multiple lines. I think spliting lines don't make sense as it makes difficult to copy & paste to execute the statement on SQL Command-Line tool.

**Comment by markstory** [comment]
@chinpei215 I did a quick update to the documentation. Feel free to add more details.

**Comment by chinpei215** [comment]
@markstory Thank you as always.

**Pull request merged**